helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/InstancePartitionsCountConstraint.java

```diff
@@ -0,0 +1,55 @@
+package org.apache.helix.controller.rebalancer.waged.constraints;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.helix.controller.rebalancer.waged.model.AssignableNode;
+import org.apache.helix.controller.rebalancer.waged.model.AssignableReplica;
+import org.apache.helix.controller.rebalancer.waged.model.ClusterContext;
+
+
+/**
+ * Evaluate by instance's current partition count and estimated max partition count
+ */
+class InstancePartitionsCountConstraint extends SoftConstraint {
+
+  InstancePartitionsCountConstraint() {
+  }
+
+  InstancePartitionsCountConstraint(float maxScore, float minScore) {
+    super(maxScore, minScore);
+  }
+
+  @Override
+  float evaluateAssignment(AssignableNode node, AssignableReplica replica, ClusterContext clusterContext) {
+    int estimatedMaxPartitionCount = clusterContext.getEstimatedMaxPartitionCount();
+    int currentPartitionCount = node.getAssignedReplicaCount();
+    // When the node is idle, return with the maxScore.
+    if (currentPartitionCount == 0) {
+      return getMaxScore();
+    }
+    //When the node usage reaches the estimated max partition, return minimal score
+    if (currentPartitionCount >= estimatedMaxPartitionCount) {
+      return getMinScore();
+    }
+
+    // Return the usage percentage by estimation
+    return (estimatedMaxPartitionCount - currentPartitionCount) / (float) estimatedMaxPartitionCount;
```

Comment:
Sorry, I am a little bit confused here. I know we are going to normalize this score later, right? Will the final result be (0.0-1.0) or (minScore, maxScore) after the normalizing? My understanding is the eventual normalized result should be (0.0 - 1.0). Please let me know if I am wrong.

Based on this understanding, it is strange that you ignore the maxScore and minScore here.
I assume something like this:
  (estimatedMaxPartitionCount - currentPartitionCount) / (float) estimatedMaxPartitionCount * (getMaxScore() - getMinScore()) + getMinScore()

Reply:
No, the min-max scaler function will guarantee you a final score within (0, 1). That's the value of a scaler function that leaves the implementation easier AS LONG AS the original score doesn't exceed the min/max and we don't have to worry about scaling and calculation ourselves.

Reply:
>  the min-max scaler function will guarantee you a final score within (0, 1)
Don't forget the scale function also needs to ensure a uniform distribution of the values.

Let's go with an example.

Max = 1000, Min = -1000
And estimatedMaxPartitionCount = 100, currentPartitionCount = 1. Then, the score returned is 0.99.
And after scale, it would be (0.99 - -1000) / 2000 = 0.500495.
And the scaled score would always be between 0.5 ~ 0.5005 (except for the 2 corner cases), right? Do you think this is the right math we want to use?

Reply:
OK. If you want uniform distribution, you can create the soft constraint with minscore = 0, maxscore = 1, these can be overridden. Correct?

Reply:
I think that is necessary. Or the result won't be comparable. And we should enforce it without providing such an option.